src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -461,19 +448,7 @@ private Array ReadArray(TypeCode kind)
             }
             else
             {
-                // custom type case
-                elementType = this.ReadTypeAfterTag();
-
-                // recursive: create instance and read elements next in stream
-                Array array = Array.CreateInstance(elementType, length);
-
-                for (int i = 0; i < length; ++i)
-                {
-                    var value = this.ReadValue();
-                    array.SetValue(value, i);
-                }
-
-                return array;
+                throw ExceptionUtilities.UnexpectedValue(elementKind);
```

Comment:
we don't support arrays of aribtrary objects.  it always needs to be an array of some type that the serializer has direct capability to serialize.  Note: arrays of arrays of arrays of ... are fine.